js/ui/windowManager.js

```diff
@@ -1832,3 +1887,509 @@ const WindowManager = new Lang.Class({
+        session.windowBuilder.get_meta_window().move_resize_frame(false,
+                                                                  rect.x,
+                                                                  rect.y,
+                                                                  rect.width,
+                                                                  rect.height);
+    },
+
+    _windowBuilderPositionChanged: function(window, session) {
+        let rect = session.windowBuilder.get_meta_window().get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.windowApp)
+            return;
+        session.windowApp.get_meta_window().move_resize_frame(false,
+                                                              rect.x,
+                                                              rect.y,
+                                                              rect.width,
+                                                              rect.height);
+    },
+
+    _windowBuilderSizeChanged: function(window, session) {
+        let rect = session.windowBuilder.get_meta_window().get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.windowApp)
+            return;
+        session.windowApp.get_meta_window().move_resize_frame(false,
+                                                              rect.x,
+                                                              rect.y,
+                                                              rect.width,
+                                                              rect.height);
+    },
+
+    _windowAppRestacked : function(overview, stackIndices, session) {
+        let focusedWindow = global.display.get_focus_window();
+        if (!focusedWindow)
+            return;
+
+        // keep track of the previous focused window so
+        // that we can show the animation accordingly
+        let previousFocused = this._previousFocusedWindow;
+        this._previousFocusedWindow = focusedWindow;
+
+        // make sure we hide the button in any other case as on
+        // top of the App and Builder window
+        session.buttonApp.hide();
+        if (session.buttonBuilder)
+            session.buttonBuilder.hide();
+
+        if (focusedWindow === session.windowApp.get_meta_window()) {
+            if (session.windowBuilder && session.windowBuilder.get_meta_window() === previousFocused) {
+                this._rotateInActors.push(session.windowApp);
+                this._rotateOutActors.push(session.windowBuilder);
+                this._animate(session.windowBuilder,
+                              session.windowApp,
+                              Gtk.DirectionType.RIGHT);
+                session.buttonApp.show();
+                session.state = SessionState.APP;
```

Comment:
I wonder if all this animation stuff (2199-2205) should really be put into a separate function. There's a lot of details here which have been repeated in a few places and it seems like missing any one of them could mess up the state.